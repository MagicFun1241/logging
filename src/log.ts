import * as stream from 'stream';
import { streamStdout, streamRaw } from './streams';
import { formatColor, formatJson, dynamicProgress } from './formatters';
import { logConfig } from './config';

const format = (level: Level, args: any[]) => {
    return (logConfig.outputJson ? formatJson : formatColor)(level, args);
};

export const log = {
    // Needs better name
    help: (...args: any[]) => {
        const output = format('HELP', args);
        return streamStdout(output);
    },
    debug: (...args: any[]) => {
        if (logConfig.outputJson || logConfig.isDebug) {
            const output = format('DEBUG', args);
            return streamStdout(output);
        }
    },
    info: (...args: any[]) => {
        const output = format('INFO', args);
        return streamStdout(output);
    },
    warn: (...args: any[]) => {
        const output = format('WARN', args);
        return streamStdout(output);
    },
    error: (...args: any[]) => {
        const output = format('ERROR', args);
        return streamStdout(output);
    },
    progress: (message: string, percentage?: number) => {
        if (logConfig.outputJson) {
            const output = formatJson('PROGRESS', [message, percentage]);
            return streamStdout(output);
        }
        return dynamicProgress.progress(message, percentage);
    },
    success: (message: string) => {
        if (logConfig.outputJson) {
            const output = formatJson('PROGRESS', [message]);
            return streamStdout(output);
        }
        return dynamicProgress.success(message);
    },
    fail: (message: string) => {
        if (logConfig.outputJson) {
            const output = formatJson('PROGRESS', [message]);
            return streamStdout(output);
        }
        return dynamicProgress.fail(message);
    },
    raw: (message: string) => {
        return streamRaw(message);
    },
    pipe: (stdout: stream.Duplex, stderr: stream.Duplex) => {
        stdout.on('data', (buffer: Buffer) => streamRaw(buffer.toString()));
        stderr.on('data', (buffer: Buffer) => streamRaw(buffer.toString()));
    }
};
