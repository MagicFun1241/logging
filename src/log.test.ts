import stripAnsi from 'strip-ansi';
import { log } from './log';

jest.mock('./streams', () => ({
    streamStdout: (x: string) => x
}));

describe('Log', () => {
    test('info', async () => {
        const results = log.debug('a', 'b');
        expect(stripAnsi(results)).toMatchSnapshot();
    });

    test('warn', async () => {
        const results = log.warn('a', 'b');
        expect(stripAnsi(results)).toMatchSnapshot();
    });

    test('error', async () => {
        const results = log.error('a', 'b');
        expect(stripAnsi(results)).toMatchSnapshot();
    });

    test('debug', async () => {
        const results = log.debug('a', 'b');
        expect(stripAnsi(results)).toMatchSnapshot();
    });
});
