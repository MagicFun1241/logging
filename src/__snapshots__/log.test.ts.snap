// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log debug 1`] = `undefined`;

exports[`Log error 1`] = `
"      · [logging/src/formatters/format-color/prefix/get-callee.ts]
        ╔═══════════════╗
        ║               ║
        ║   > ERROR <   ║
        ║               ║
        ║   a b         ║
        ║               ║
        ╚═══════════════╝"
`;

exports[`Log info 1`] = `undefined`;

exports[`Log warn 1`] = `
"      · [logging/src/formatters/format-color/prefix/get-callee.ts]
        ┌─────────────────┐
        │                 │
        │   > WARNING <   │
        │                 │
        │   a b           │
        │                 │
        └─────────────────┘"
`;
